allennlp/data/dataset_readers/reading_comprehension/util.py

```diff
@@ -213,3 +213,136 @@ def make_reading_comprehension_instance(question_tokens: List[Token],
+    in a reading comprehension model.
+
+    Creates an ``Instance`` with at least these fields: ``question`` and ``passage``, both
+    ``TextFields``; and ``metadata``, a ``MetadataField``.  Additionally, if both ``answer_texts``
+    and ``char_span_starts`` are given, the ``Instance`` has ``span_start`` and ``span_end``
+    fields, which are both ``IndexFields``.
+
+    Parameters
+    ----------
+    question_list_tokens : ``List[List[Token]]``
+        An already-tokenized question.
+    passage_tokens : ``List[Token]``
+        An already-tokenized passage that contains the answer to the given question.
+    token_indexers : ``Dict[str, TokenIndexer]``
+        Determines how the question and passage ``TextFields`` will be converted into tensors that
+        get input to a model.  See :class:`TokenIndexer`.
+    passage_text : ``str``
+        The original passage text.  We need this so that we can recover the actual span from the
+        original passage that the model predicts as the answer to the question.  This is used in
+        official evaluation scripts.
+    token_spans_lists : ``List[List[Tuple[int, int]]]``, optional
+        Indices into ``passage_tokens`` to use as the answer to the question for training.  This is
+        a list because there might be several possible correct answer spans in the passage.
+        Currently, we just select the most frequent span in this list (i.e., SQuAD has multiple
+        annotations on the dev set; this will select the span that the most annotators gave as
+        correct).
+    yesno_list : List[int]
+    followup_list : List[int]
+    prev_a : int
+    additional_metadata : ``Dict[str, Any]``, optional
+        The constructed ``metadata`` field will by default contain ``original_passage``,
+        ``token_offsets``, ``question_tokens``, ``passage_tokens``, and ``answer_texts`` keys.  If
+        you want any other metadata to be associated with each instance, you can pass that in here.
+        This dictionary will get added to the ``metadata`` dictionary we already construct.
+    """
+    additional_metadata = additional_metadata or {}
+    fields: Dict[str, Field] = {}
+    passage_offsets = [(token.idx, token.idx + len(token.text)) for token in passage_tokens]
+    # This is separate so we can reference it later with a known type.
+    passage_field = TextField(passage_tokens, token_indexers)
+    fields['passage'] = passage_field
+    fields['question'] = ListField([TextField(q_tokens, token_indexers) for q_tokens in question_list_tokens])
+    metadata = {'original_passage': passage_text,
+                'token_offsets': passage_offsets,
+                'question_tokens': [[token.text for token in q] for q in question_list_tokens],
+                'passage_tokens': [token.text for token in passage_tokens],
+                }
+    p1_answer_marker_list = []
+    p2_answer_marker_list = []
+    p3_answer_marker_list = []
+
+    def get_tag(i, i_name):
```

Comment:
Please add type hints and descriptions to these functions - why are we wrapping the a piece of text and an index in brackets?